src/dataflow/src/source/postgres.rs

```diff
@@ -197,22 +190,69 @@ impl PostgresSourceReader {
                 .await?;
             for msg in data {
                 if let SimpleQueryMessage::Row(row) = msg {
-                    let mut mz_row = Row::default();
-                    let rel_id: Datum = (info.rel_id as i32).into();
-                    mz_row.push(rel_id);
-                    mz_row.push_list((0..row.len()).map(|n| {
-                        let a: Datum = row.get(n).into();
-                        a
-                    }));
+                    let mz_row = try_fatal!(self.row_from_query_row(info.rel_id, row));
                     try_recoverable!(snapshot_tx.insert(mz_row.clone()).await);
                     try_fatal!(buffer.write(&try_fatal!(bincode::serialize(&mz_row))).await);
                 }
             }
         }
         client.simple_query("COMMIT;").await?;
+        for table in publication_tables {
+            self.synced_rel_ids.insert(table.rel_id);
+        }
         Ok(())
     }
 
+    /// Takes a snapshot of the specified relation, syncing the data to the source.
+    /// This function will be called when we detect a new relation has been added to the
+    /// Postgres publication.
+    async fn produce_snapshot(&mut self, rel_id: u32) -> Result<Vec<Row>, ReplicationError> {
```

Comment:
There is definitely redundancy between this function and `produce_initial_snapshot`. I made them separate to avoid putting a bunch of `if/elses` in.